### ISSUE TITLE
 Unknown variable 'R2_Files' -- Make sure it is not misspelt and defined somewhere in the script before using it

### ISSUE THREAD

**Issue by Xinming-W** [RenneLab/CnR-flow#1]
hello, when I use CnR-flow for one sample with treat file but no ctrl file. It detected the input,but said Unknown variable 'R2_Files'.
My command is :
`nextflow run CnR-flow --mode run`
The output is:
"N E X T F L O W  ~  version 20.10.0
Launching `RenneLab/CnR-flow` [golden_euler] - revision: aa7cf33e33 [master]

Utilizing Run Mode: run

-- Preparing Workflow Environment --

2 Input Files Detected.

-- Executing Workflow --

[-        ] process > CnR_S0_A_GetSeqLen   -
[-        ] process > CnR_S0_B_MergeFastqs -
[-        ] process > CnR_S0_C_FastQCPre   -
[-        ] process > CnR_S1_A_Trim        -
[-        ] process > CnR_S1_B_Retrim      -
[-        ] process > CnR_S1_C_FastQCPost  -
[-        ] process > CnR_S2_A_Aln_Ref     -
[-        ] process > CnR_S2_B_Modify_Aln  -
Error executing process > 'CnR_S0_B_MergeFastqs (Aged_HSC_CTCF_10)'

Caused by:
  Unknown variable 'R2_Files' -- Make sure it is not misspelt and defined somewhere in the script before using it

Source block:
  run_id = "${task.tag}.${task.process}"
  out_log_name = "${run_id}.nf.log.txt"
  merge_fastqs_dir = "${params.merge_fastqs_dir}"
  R1_files = fastq.findAll {fn ->
                  "${fn}".contains("_R1_") || "${fn}".contains("_1.f") || "${fn}".contains("_1_")
              }
  R2_files = fastq.findAll {fn -> 
                  "${fn}".contains("_R2_") || "${fn}".contains("_2.f") || "${fn}".contains("_2_")
              }
  R1_out_file = "${params.merge_fastqs_dir}/${name}_R1_001.fastq.gz"
  R2_out_file = "${params.merge_fastqs_dir}/${name}_R2_001.fastq.gz"
  if( R1_files.size() < 1 || R2_files.size() < 1 ) {
                  message = "Merge Error:\nR1 Files: ${R1_files}\nR2 Files: ${R2_Files}"
                  throw new Exception(message)
              }
  check_command = ""
  if( "${R1_files[0]}".endsWith('.gz') ) {
                  check_command = "gzip -vt ${R1_files.join(' ')} ${R2_files.join(' ')}"
              }
  if( R1_files.size() == 1 && R2_files.size() == 1 ) {
                  command = '''
                  echo "No Merge Necessary. Renaming Files..."
                  mkdir !{merge_fastqs_dir}
  
                  # Check File Integrity if gzipped
                  !{check_command}
  
                  set -v -H -o history
                  mv -v "!{R1_files[0]}" "!{R1_out_file}"
                  mv -v "!{R2_files[0]}" "!{R2_out_file}"
                  set +v +H +o history
                  R1_OUT_LEN=$(zcat -f !{R1_out_file} | wc -l)
                  R2_OUT_LEN=$(zcat -f !{R2_out_file} | wc -l)
                  echo "R1 Lines: ${R1_OUT_LEN}"
                  echo "R2 Lines: ${R2_OUT_LEN}"
                  if [ "${R1_OUT_LEN}" == "0" || "${R2_OUT_LEN}" == "0" ]; then
                      echo "Input file of zero length detected."
                      exit 1
                  fi
                  '''
              } else {
                  command = '''
                  mkdir !{merge_fastqs_dir}
[-        ] process > CnR_S0_A_GetSeqLen   [  0%] 0 of 1
[-        ] process > CnR_S0_B_MergeFastqs -
[-        ] process > CnR_S0_C_FastQCPre   -
[-        ] process > CnR_S1_A_Trim        -
[-        ] process > CnR_S1_B_Retrim      -
[-        ] process > CnR_S1_C_FastQCPost  -
[-        ] process > CnR_S2_A_Aln_Ref     -
[-        ] process > CnR_S2_B_Modify_Aln  -
[-        ] process > CnR_S2_C_Make_Bdg    -
[-        ] process > CnR_S3_A_Aln_Spike   -
[-        ] process > CnR_S3_B_Norm_Bdg    -
[-        ] process > CnR_S4_A_Make_BigWig -
[-        ] process > CnR_S5_A_Peaks_MACS  -
[-        ] process > CnR_S5_B_Peaks_SEACR -
Error executing process > 'CnR_S0_B_MergeFastqs (Aged_HSC_CTCF_10)'

Caused by:
  Unknown variable 'R2_Files' -- Make sure it is not misspelt and defined somewhere in the script before using it

Source block:
  run_id = "${task.tag}.${task.process}"
  out_log_name = "${run_id}.nf.log.txt"
  merge_fastqs_dir = "${params.merge_fastqs_dir}"
  R1_files = fastq.findAll {fn ->
                  "${fn}".contains("_R1_") || "${fn}".contains("_1.f") || "${fn}".contains("_1_")
              }
  R2_files = fastq.findAll {fn -> 
                  "${fn}".contains("_R2_") || "${fn}".contains("_2.f") || "${fn}".contains("_2_")
              }
  R1_out_file = "${params.merge_fastqs_dir}/${name}_R1_001.fastq.gz"
  R2_out_file = "${params.merge_fastqs_dir}/${name}_R2_001.fastq.gz"
  if( R1_files.size() < 1 || R2_files.size() < 1 ) {
                  message = "Merge Error:\nR1 Files: ${R1_files}\nR2 Files: ${R2_Files}"
                  throw new Exception(message)
              }
  check_command = ""
  if( "${R1_files[0]}".endsWith('.gz') ) {
                  check_command = "gzip -vt ${R1_files.join(' ')} ${R2_files.join(' ')}"
              }
  if( R1_files.size() == 1 && R2_files.size() == 1 ) {
                  command = '''
                  echo "No Merge Necessary. Renaming Files..."
                  mkdir !{merge_fastqs_dir}
  
                  # Check File Integrity if gzipped
                  !{check_command}
  
                  set -v -H -o history
                  mv -v "!{R1_files[0]}" "!{R1_out_file}"
                  mv -v "!{R2_files[0]}" "!{R2_out_file}"
                  set +v +H +o history
                  R1_OUT_LEN=$(zcat -f !{R1_out_file} | wc -l)
                  R2_OUT_LEN=$(zcat -f !{R2_out_file} | wc -l)
                  echo "R1 Lines: ${R1_OUT_LEN}"
                  echo "R2 Lines: ${R2_OUT_LEN}"
                  if [ "${R1_OUT_LEN}" == "0" || "${R2_OUT_LEN}" == "0" ]; then
                      echo "Input file of zero length detected."
                      exit 1
                  fi
                  '''
              } else {
                  command = '''
                  mkdir !{merge_fastqs_dir}
                  mkdir !{merge_fastqs_dir}
 
                  # Check File Integrity if gzipped
                  !{check_command}
 
                  echo -e "\\nCombining Files: !{R1_files.join(' ')}"
                  echo "    Into: !{R1_out_file}"
                  set -v -H -o history
                  cat '!{R1_files.join("' '")}' > '!{R1_out_file}'
                  set +v +H +o history
 
                  echo -e "\\nCombining Files: !{R2_files.join(' ')}"
                  echo "    Into: !{R2_out_file}"
                  set -v -H -o history
                  cat '!{R2_files.join("' '")}' > '!{R2_out_file}'
                  set +v +H +o history
 
                  R1_OUT_LEN=$(zcat -f !{R1_out_file} | wc -l)
                  R2_OUT_LEN=$(zcat -f !{R2_out_file} | wc -l)
                  echo "R1 Lines: ${R1_OUT_LEN}"
                  echo "R2 Lines: ${R2_OUT_LEN}"
                  if [ "${R1_OUT_LEN}" == "0" || "${R2_OUT_LEN}" == "0" ]; then
                      echo "Input file of zero length detected."
                      exit 1
                  fi
                  '''
              }
  command

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
"
So what's wrong?
Thanks.

**Comment by dstrib** [comment]
Thanks for the question, and for posting the detailed problem message.
It is trying to send you the error message in this block:
```
if( R1_files.size() < 1 || R2_files.size() < 1 ) {
message = "Merge Error:\nR1 Files: ${R1_files}\nR2 Files: ${R2_Files}"
throw new Exception(message)
}
```
I will push a fix to this, but the then it still should (correctly) give you an error message saying that it can't identify automatically which items are your "R1" files and which are your "R2" files (to group the correct files accordingly.

1. Are the sequences you're providing paired-end?
2. If yes to 1., can you provide a brief example of the way they are named? "_1" and "_2", etc? 
3. If you would like to skip the per-sample lane merging step which requires this setting, you can set "do_merge_lanes = false" in your "nextflow.config" file.

**Comment by Xinming-W** [comment]
Thanks for your prompt reply.I set "do_merge_lanes = false" in the "nextflow.config" file and it seems to be working properly now.
![image](https://user-images.githubusercontent.com/43260022/105655717-97b5ac00-5efb-11eb-9751-8148c4668de9.png)

**Comment by dstrib** [comment]
Great! Also, would you mind providing an example of the way your files are named? I want to make sure that naming scheme can be detected, or see if I need to add a parameter for custom detection.

**Comment by Xinming-W** [comment]
OK, I tried two samples:
the first one is `Aged_HSC_CTCF_10_R1_001.fastq.gz,Aged_HSC_CTCF_10_R2_002.fastq.gz`, I set `treat_fastqs=['Aged_HSC_CTCF_10_R{1,2}*.fastq.gz']`. It can be detected.
the second one is `Aged_MPP_CTCF_11_R1.fq.gz,Aged_MPP_CTCF_11_R2.fq.gz`, I set `treat_fastqs=Aged_MPP_CTCF_11_R{1,2}*.fq.gz`.  It can also be detected.

**Comment by Xinming-W** [comment]
I don't have different lane fastq files, so can't test per-sample lane merging step.

**Comment by dstrib** [comment]
Thanks very much for posting this! Ideally the "merging" step will just passthrough any samples that don't need merged, instead of giving you the error like it did.
I see that I hadn't included the possibility of files with "_R1" in the name (as opposed to "_R1_"), so I can now fix that so you should be able to run with "do_merge_lanes" enabled (even though in your case it will just be ignored). Thanks!

**Comment by dstrib** [comment]
Fixed with #4c05f4832d6134fef0c55cf99b3703b41eb3d227

**Comment by Xinming-W** [comment]
Thanks too, good job！
